src/heap/sharedMem.ml

```diff
@@ -1071,6 +1111,298 @@ module NewAPI = struct
     else
       Some addr
 
+  (** Field utils *)
+
+  let set_generic offset base = unsafe_write_addr_at (get_heap ()) (offset base)
+
+  let get_generic offset base = read_addr (get_heap ()) (offset base)
+
+  let get_generic_opt offset base = read_opt (get_generic offset base)
+
+  let get_generic_int offset base = read_tagged_int (get_heap ()) (offset base)
+
+  (** Skip lists
+   *
+   * We can use skip lists to store sets of values supporting concurrent search,
+   * insertion, and deletion in log(n) time.
+   *
+   * Insertion and deletion operations are lock-free, search is wait-free.
+   *
+   * This is a "textbook" implementation taken from Herlihy et al's "The Art of
+   * Multiprocessor Programming" 2nd Edition, section 14.4. It builds on the
+   * implementation from the Multicore OCaml runtime. This implementation
+   * differs from those implementations in the following ways:
+   *
+   * 1) The `preds` array populated by `sklist_find` does not contain pointers
+   *    to the predecessor node itself, but pointers to predecessor node's
+   *    forward pointer at the given level.
+   * 2) There are no sentinel head/tail nodes. We don't need the head node due
+   *    to (1) since preds can contain contain pointers to the list object's
+   *    forward pointers. Instead of a tail node, we use NULL_ADDR to indicate
+   *    the end of a list.
+   * 3) We rely on sharedmem GC to reclaim space from deleted nodes
+   *
+   *)
+
+  (* To provide log(n) operations, we need enough levels depending on the
+   * expected number of elements in the set. Given p=1/4, capping the number of
+   * levels at 12 lets us store 16,777,216 (4^12) elements. *)
+  let sklist_num_levels = 12
+
+  (* Skip lists are probabilistic. Each node is created with a random height,
+   * given by this function. We take p=1/4, meaning that 3/4 of nodes will have
+   * height 1, 3/16 will have height 1, 3/64 will have height 2, and so on. *)
+  let sklist_random_level () =
+    let num_rolls = sklist_num_levels - 1 in
+    (* Random.int only gives 30 bits of randomness and we use 2 bits per roll *)
+    assert (num_rolls <= 15);
+    let mask = (1 lsl (2 * num_rolls)) - 1 in
+    let r = ref (Random.bits () land mask) in
+    let l = ref 0 in
+    while !r land 3 == 3 do
+      (* two random bits will be set with probability 1/4 *)
+      incr l;
```

Comment:
There is an interesting note on the ocaml implementation that " the least significant bits are
   "less random" than the most significant bits with a modulus of 2^m"

Reply:
This is true for the LCG generator that they're using, but the OCaml Random module uses an algorithm that provides better mixing.